### ISSUE TITLE
NullPointerException, NotSerializableException with 0.7 and 0.6.2 with jenkins 2.80, 2.81, 2.82, 2.83

### ISSUE THREAD

**Issue by ThStock** [jenkinsci/docker-plugin#519]
- there are no old configuration files
 - docker engine: Version = 17.03.0-ce, API Version = 1.26
 - docker-plugin version 0.17
 - Jenkins ver. 2.80
 - Jenkins ver. 2.81
 - Jenkins ver. 2.82
```
13:44:44 java.lang.NullPointerException
13:44:44 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.buildImage(DockerBuilderPublisher.java:331)
13:44:44 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.run(DockerBuilderPublisher.java:286)
13:44:44 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.perform(DockerBuilderPublisher.java:429)
13:44:44 	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
13:44:44 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
13:44:44 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:736)
13:44:44 	at hudson.model.Build$BuildExecution.build(Build.java:206)
13:44:44 	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
13:44:44 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:496)
13:44:44 	at hudson.model.Run.execute(Run.java:1724)
13:44:44 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
13:44:44 	at hudson.model.ResourceController.execute(ResourceController.java:97)
13:44:44 	at hudson.model.Executor.run(Executor.java:419)
```
 - Jenkins ver. 2.83
```
13:25:44 FATAL: remote file operation failed: /../.. at hudson.remoting.Channel@28cc696c:JNLP4-connect connection from ....: java.io.IOException: Unable to serialize hudson.FilePath$FileCallableWrapper@2b9fcb57
13:25:44 java.io.NotSerializableException: org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint
13:25:44 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
13:25:44 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
13:25:44 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
13:25:44 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
13:25:44 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
13:25:44 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
13:25:44 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
13:25:44 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
13:25:44 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
13:25:44 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
13:25:44 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
13:25:44 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
13:25:44 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
13:25:44 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
13:25:44 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
13:25:44 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
13:25:44 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
13:25:44 	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
13:25:44 	at hudson.remoting.UserRequest._serialize(UserRequest.java:245)
13:25:44 	at hudson.remoting.UserRequest.serialize(UserRequest.java:254)
13:25:44 Caused: java.io.IOException: Unable to serialize hudson.FilePath$FileCallableWrapper@2b9fcb57
13:25:44 	at hudson.remoting.UserRequest.serialize(UserRequest.java:256)
13:25:44 	at hudson.remoting.UserRequest.<init>(UserRequest.java:97)
13:25:44 	at hudson.remoting.Channel.call(Channel.java:902)
13:25:44 	at hudson.FilePath.act(FilePath.java:987)
13:25:44 Caused: java.io.IOException: remote file operation failed: /../.. at hudson.remoting.Channel@28cc696c:JNLP4-connect connection from ...
13:25:44 	at hudson.FilePath.act(FilePath.java:994)
13:25:44 	at hudson.FilePath.act(FilePath.java:976)
13:25:44 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.buildImage(DockerBuilderPublisher.java:339)
13:25:44 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.run(DockerBuilderPublisher.java:286)
13:25:44 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.perform(DockerBuilderPublisher.java:429)
13:25:44 	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
13:25:44 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
13:25:44 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:736)
13:25:44 	at hudson.model.Build$BuildExecution.build(Build.java:206)
13:25:44 	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
13:25:44 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:496)
13:25:44 	at hudson.model.Run.execute(Run.java:1724)
13:25:44 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
13:25:44 	at hudson.model.ResourceController.execute(ResourceController.java:97)
13:25:44 	at hudson.model.Executor.run(Executor.java:419)
```
 - docker-plugin version 0.16.2
 - Jenkins ver. 2.80
```
14:04:58 java.lang.NullPointerException
14:04:58 	at java.net.URI$Parser.parse(URI.java:3042)
14:04:58 	at java.net.URI.<init>(URI.java:588)
14:04:58 	at java.net.URI.create(URI.java:850)
14:04:58 	at com.nirima.jenkins.plugins.docker.client.ClientConfigBuilderForPlugin.forServer(ClientConfigBuilderForPlugin.java:70)
14:04:58 	at com.nirima.jenkins.plugins.docker.client.ClientConfigBuilderForPlugin.forCloud(ClientConfigBuilderForPlugin.java:56)
14:04:58 	at com.nirima.jenkins.plugins.docker.DockerCloud.getClient(DockerCloud.java:198)
14:04:58 	at com.nirima.jenkins.plugins.docker.DockerCloud.isTriton(DockerCloud.java:659)
14:04:58 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.getCloud(DockerBuilderPublisher.java:154)
14:04:58 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.perform(DockerBuilderPublisher.java:361)
14:04:58 	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
14:04:58 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
14:04:58 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:736)
14:04:58 	at hudson.model.Build$BuildExecution.build(Build.java:206)
14:04:58 	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
14:04:58 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:496)
14:04:58 	at hudson.model.Run.execute(Run.java:1724)
14:04:58 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
14:04:58 	at hudson.model.ResourceController.execute(ResourceController.java:97)
14:04:58 	at hudson.model.Executor.run(Executor.java:419)
```

**Comment by ndeloof** [comment]
thanks for reporting this. 
May I suggest you use https://wiki.jenkins.io/display/JENKINS/CloudBees+Docker+Build+and+Publish+plugin to build and publish docker images ?

**Comment by ThStock** [comment]
Is docker-plugin deprecated and should be replaced by cloudbees plugin?

**Comment by ndeloof** [comment]
@ThStock definitively not, but this build step to "build and publish docker image" is not it's core feature, and as a dedicated plugin exists better to avoid too much responsibilities within the same plugin. Maybe I should implement some automated configuration migration

**Comment by ndeloof** [comment]
hum, as `com.github.dockerjava.core.DockerClientImpl` isn't serialisable I even wonder this has even been working :-\

**Comment by ndeloof** [comment]
seems our last chance is to only compute the context tar stream computation on agent, then run the `docker build` client on master.

**Comment by ThStock** [comment]
I've tried the CloudBees plugin; it doesn't work anymore ... latest release was Oct 2016

**Comment by ndeloof** [comment]
ok, thanks for reporting this.

**Comment by ThStock** [comment]
a new exception if I try to enable "push image"

```
org.kohsuke.stapler.NoStaplerConstructorException: There's no @DataBoundConstructor on any constructor of boolean
	at org.kohsuke.stapler.ClassDescriptor.loadConstructorParamNames(ClassDescriptor.java:265)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:764)
	at org.kohsuke.stapler.RequestImpl.access$200(RequestImpl.java:83)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:678)
Caused: java.lang.IllegalArgumentException: Failed to instantiate boolean from {"registry":{"url":"my-repo","credentialsId":""}}
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:680)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:478)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:777)
Caused: java.lang.IllegalArgumentException: Failed to convert the pushOnSuccess parameter of the constructor public com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher(java.lang.String,org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint,java.lang.String,java.lang.String,boolean,org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint,boolean,boolean)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:779)
	at org.kohsuke.stapler.RequestImpl.access$200(RequestImpl.java:83)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:678)
Caused: java.lang.IllegalArgumentException: Failed to instantiate class com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher from {"dockerFileDirectory":"","fromRegistry":{"url":"","credentialsId":""},"cloud":"anycloud","tagsString":"my-image/of/some:latest","pushOnSuccess":{"registry":{"url":"my-repo","credentialsId":""}},"cleanImages":false,"cleanupWithJenkinsJobDelete":false,"stapler-class":"com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher","$class":"com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher"}
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:680)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:478)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:474)
	at hudson.model.Descriptor.newInstance(Descriptor.java:592)
Caused: java.lang.Error: Failed to instantiate class com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher from {"dockerFileDirectory":"","fromRegistry":{"url":"","credentialsId":""},"cloud":"anycloud","tagsString":"my-image/of/some:latest","pushOnSuccess":{"registry":{"url":"my-repo","credentialsId":""}},"cleanImages":false,"cleanupWithJenkinsJobDelete":false,"stapler-class":"com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher","$class":"com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher"}
	at hudson.model.Descriptor.newInstance(Descriptor.java:600)
	at hudson.model.Descriptor.newInstancesFromHeteroList(Descriptor.java:1048)
	at hudson.model.Descriptor.newInstancesFromHeteroList(Descriptor.java:1010)
	at hudson.util.DescribableList.rebuildHetero(DescribableList.java:208)
```
hmm no more docker image creation :worried:

**Comment by ndeloof** [comment]
this one I already noticed. Will need to fully review this builder

**Comment by ThStock** [comment]
I tried Jenkins 2.84 with docker-plugin 0.18

```
17:09:05 FATAL: remote file operation failed: /... at hudson.remoting.Channel@67427469:JNLP4-connect connection from 10.10.10.10/10.10.10.10:57745: java.io.IOException: Unable to serialize hudson.FilePath$FileCallableWrapper@26d8fdfc
17:09:05 java.io.NotSerializableException: org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint
17:09:05 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
17:09:05 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
17:09:05 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
17:09:05 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
17:09:05 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
17:09:05 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
17:09:05 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
17:09:05 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
17:09:05 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
17:09:05 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
17:09:05 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
17:09:05 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
17:09:05 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
17:09:05 	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
17:09:05 	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
17:09:05 	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
17:09:05 	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
17:09:05 	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
17:09:05 	at hudson.remoting.UserRequest._serialize(UserRequest.java:245)
17:09:05 	at hudson.remoting.UserRequest.serialize(UserRequest.java:254)
17:09:05 Caused: java.io.IOException: Unable to serialize hudson.FilePath$FileCallableWrapper@26d8fdfc
17:09:05 	at hudson.remoting.UserRequest.serialize(UserRequest.java:256)
17:09:05 	at hudson.remoting.UserRequest.<init>(UserRequest.java:97)
17:09:05 	at hudson.remoting.Channel.call(Channel.java:902)
17:09:05 	at hudson.FilePath.act(FilePath.java:987)
17:09:05 Caused: java.io.IOException: remote file operation failed: /... at hudson.remoting.Channel@67427469:JNLP4-connect connection from 10.10.10.10/10.10.10.10:57745
17:09:05 	at hudson.FilePath.act(FilePath.java:994)
17:09:05 	at hudson.FilePath.act(FilePath.java:976)
17:09:05 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.buildImage(DockerBuilderPublisher.java:339)
17:09:05 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.run(DockerBuilderPublisher.java:286)
17:09:05 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.perform(DockerBuilderPublisher.java:429)
17:09:05 	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
17:09:05 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
17:09:05 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:736)
17:09:05 	at hudson.model.Build$BuildExecution.build(Build.java:206)
17:09:05 	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
17:09:05 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:496)
17:09:05 	at hudson.model.Run.execute(Run.java:1724)
17:09:05 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
17:09:05 	at hudson.model.ResourceController.execute(ResourceController.java:97)
17:09:05 	at hudson.model.Executor.run(Executor.java:421)
```

**Comment by ndeloof** [comment]
7f8157b is not part of a release yet. If you want to confirm the fix, you'll need to build a SNAPSHOT and install manually

**Comment by ThStock** [comment]
I tried it, but can't resolve all dependecies (org.netbeans.modules:org-netbeans-insane:jar:RELEASE72) ... I'll wait for the release

**Comment by ndeloof** [comment]
I guess you use an internal repository mirror, plugin dependencies are all available on jenkins.io infra as configured by plugin parent pom.xml (http://repo.jenkins-ci.org/public/)

A simpler option is for you to download the artifact built by CI : https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fdocker-plugin/detail/master/24/artifacts

**Comment by ThStock** [comment]
the exceptions are fixed ... but push to remote throws the following (I've defined credentials for the build)
```
14:07:46 com.github.dockerjava.api.exception.DockerClientException: Could not push image: no basic auth credentials
14:07:46 	at com.github.dockerjava.core.command.PushImageResultCallback.awaitSuccess(PushImageResultCallback.java:49)
14:07:46 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.pushImages(DockerBuilderPublisher.java:405)
14:07:46 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher$Run.run(DockerBuilderPublisher.java:303)
14:07:46 	at com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.perform(DockerBuilderPublisher.java:434)
```

**Comment by ndeloof** [comment]
ok, I made a stupid mistake and just fixed it. Thanks for you patience

**Comment by ThStock** [comment]
do you haven an hpi download on ci.jenkins.io for me?

**Comment by ndeloof** [comment]
sure, you can always get the latest bits form master branch on https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fdocker-plugin/activity
(last build failure on Linux node is caused by an infra issue)

**Comment by ThStock** [comment]
Now all exceptions are gone, but pushing doesn't work
```
16:47:58 Removing intermediate container 6aa7c82dc1f3
16:47:58
16:47:58 Successfully built 4b81ca7c79cd
16:47:58
16:47:58 Tagging built image with my.repo-com/any/any:latest
16:47:58 Docker Build Response : 4b81ca7c79cd
16:47:58 Pushing [my.repo-com/any/any:latest]
16:47:58 Docker Build Done
16:47:58 The push refers to a repository [my.repo-com/any/any]
16:58:26 4c45c8e71b4a: Preparing
16:58:26 d1ac57eeb246: Preparing
16:58:26 796d69313fd5: Preparing
16:58:26 07518bbe7c43: Preparing
16:58:26 f8faacf8139c: Preparing
16:58:26 7aa57d3f4404: Preparing
16:58:26 ce8f2ebb2511: Preparing
16:58:26 b0f743408169: Preparing
16:58:26 5bef08742407: Preparing
16:58:26 7aa57d3f4404: Waiting
16:58:26 ce8f2ebb2511: Waiting
16:58:26 b0f743408169: Waiting
16:58:26 5bef08742407: Waiting
16:58:26 Finished: SUCCESS
```

**Comment by ThStock** [comment]
I'll offer my help to fix this problem :simple_smile:

**Comment by ndeloof** [comment]
I ran some tests with a private registry on dockerhub and discovered another issue with push : token authentication doesn't work as expected, it only works with username+password (which means ECR / GCR won't work). I'm investigating for a fix 💻

**Comment by ndeloof** [comment]
fixed in 1.0
